### DISCUSSION TITLE
Invalid AST Node

### DISCUSSION THREAD

**Discussion by narayanpromax** [urql-graphql/urql#3090] [Q&A] [answered]
I'm dynamically building query and variables using gql-query-builder and pass to client.query but it gives me this error

**Case 1:**
`const { data, fetching, error } = await client.query({
    query: `${query}`,
    variables
  })
  .toPromise()
  .then(result=>{
    console.log('Result: ', result)
    return result
  })`
  
  **Case 2:**
 `const { data, fetching, error } = await client.query({
    query: gql`${query}`,
    variables
  })
  .toPromise()
  .then(result=>{
    console.log('Result: ', result)
    return result
  })`
  
   **Case 3:**
 `const { data, fetching, error } = await client.query({
    query,
    variables
  })
  .toPromise()
  .then(result=>{
    console.log('Result: ', result)
    return result
  })`
  
  All three fail with sort of below error. I think some type inference is the problem between String and DocumentNode but not able to fix it.

`Error: Invalid AST Node: { query: "query ($limit: Int, $offset: Int, $order_by: [orders_order_by!], $where: orders_bool_exp) { orders (limit: $limit, offset: $offset, order_by: $order_by, where: $where) { id } orders_aggregate (where: $where) { aggregate { count(distinct: true, columns: id) } } }", variables: { limit: 10, offset: 0, order_by: {}, where: {} } }`

**Comment by kitten** [comment, marked as answer]
That's not a query document, that's an entire request, so you'll have to pass in `...query` in your above code example

**Reply by narayanpromax** [reply]
Could you please write in code snippet where exactly?

**Reply by kitten** [reply]
Just to clarify, the error states what you need to do. You passed `{ query: "...", variables: {} }` into `query`, but query needs to be the document node or query string.

You're also saying you're passing `await client.query({ query: query, variables })`
although, as the docs state `client.query` accepts `client.query(query, variables)`

So you need to pass, assuming `query` is set to your `gql-query-builder` and has a shape of `{ query, variables }`: `client.query(query.query, query.variables)`.

**Reply by narayanpromax** [reply]
**I have below code as in Case 3 above. Same Error. Do you want content of Maker Query and Maker variables?**

`const { query, variables } = generateQuery(resource, sort, filters, pagination, metaData)

console.log('Maker Query: ', query);

console.log('Maker Variables : ', variables); 

const { data, fetching, error } = await client.query({
    query,
    variables
  })
  .toPromise()
  .then(result=>{
    console.log('Result: ', result)
    return result
  })`
  
**Maker Query and Variables can be seen in below screenshot:** 
  
![screenshot-nimbusweb me-2023 03 23-23_35_38](https://user-images.githubusercontent.com/61919250/227304808-6b805f24-0b43-42eb-b9a1-2346447a1b3e.png)

**Reply by narayanpromax** [reply]
FurtherMore i copied same query as in console of Maker Query and made a query out of it, copied variables from Maker Variables in console and made a const of it.

1) Tried useQuery in JSX (gives results and works)
2) Tried client.query in JS as in above but with query: GET_ORDERS (copied from console content), gives AST node error

**Reply by JoviDeCroock** [reply]
Read his reply 😅 `client.query(query, variables)` not as an object your examples are doing `client.query({ query, variables })`

**Reply by narayanpromax** [reply]
Thanks @JoviDeCroock and @kitten. Silly mistake. Since it was code in your reply my eyes auto added {} 😅. I went to docs and saw and removed {} and tried and co-incidentally when i fixed it @JoviDeCroock's reply came
